crates/sui-core/src/rest_index.rs

```diff
@@ -0,0 +1,509 @@
+use sui_types::object::Object;
+use sui_types::object::Owner;
+use sui_types::storage::error::Error as StorageError;
+use sui_types::type_resolver::LayoutResolver;
+use tracing::{debug, info};
+use typed_store::rocks::{DBMap, MetricConf};
+use typed_store::traits::Map;
+use typed_store::traits::{TableSummary, TypedStoreDebug};
+use typed_store::TypedStoreError;
+use typed_store_derive::DBMapUtils;
+
+#[derive(Clone, Copy, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord, Debug)]
+struct OwnerIndexKey {
+    owner: SuiAddress,
+    object_id: ObjectID,
+}
+
+impl OwnerIndexKey {
+    fn new(owner: SuiAddress, object_id: ObjectID) -> Self {
+        Self { owner, object_id }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct OwnerIndexInfo {
+    // object_id of the object is a part of the Key
+    pub version: SequenceNumber,
+    pub type_: MoveObjectType,
+}
+
+impl OwnerIndexInfo {
+    pub fn new(object: &Object) -> Self {
+        Self {
+            version: object.version(),
+            type_: object.type_().expect("packages cannot be owned").to_owned(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord, Debug)]
+struct DynamicFieldKey {
+    parent: ObjectID,
+    field_id: ObjectID,
+}
+
+impl DynamicFieldKey {
+    fn new<P: Into<ObjectID>>(parent: P, field_id: ObjectID) -> Self {
+        Self {
+            parent: parent.into(),
+            field_id,
+        }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct DynamicFieldIndexInfo {
+    // field_id of this dynamic field is a part of the Key
+    pub dynamic_field_type: DynamicFieldType,
+    pub name_type: TypeTag,
+    pub name_value: Vec<u8>,
+    // TODO do we want to also store the type of the value? We can get this for free for
+    // DynamicFields, but for DynamicObjects it would require a lookup in the DB on init, or
+    // scanning the transaction's output objects for the coorisponding Object to retreive its type
+    // information.
+    //
+    // pub value_type: TypeTag,
+    /// ObjectId of the child object when `dynamic_field_type == DynamicFieldType::DynamicObject`
+    pub dynamic_object_id: Option<ObjectID>,
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize, Eq, PartialEq, Debug)]
+pub struct TransactionInfo {
+    checkpoint: u64,
+    timestamp_ms: u64,
+}
+
+/// RocksDB tables for the RestIndexStore
+///
+/// NOTE: Authors and Reviewers before adding any new tables ensure that they are either:
+/// - bounded in size by the live object set
+/// - are prune-able and have corresponding logic in the `prune` function
+#[derive(DBMapUtils)]
+struct IndexStoreTables {
+    /// An index of extra metadata for Transactions.
+    ///
+    /// Only contains entries for transactions which have yet to be pruned from the main database.
+    transactions: DBMap<TransactionDigest, TransactionInfo>,
```

Comment:
any reason we need a separate implementation of this table for REST API?
* `timestamps` is a [deprecated](https://github.com/MystenLabs/sui/blob/0211924835a1ddfd40ad95e3e371609787ebc370/crates/sui-storage/src/indexes.rs#L176) column family in JSON RPC index store  
* prunable mapping to checkpoint already [exists](https://github.com/MystenLabs/sui/blob/0211924835a1ddfd40ad95e3e371609787ebc370/crates/sui-core/src/authority/authority_store_tables.rs#L97) in perpetual db

Reply:
So i think that `timestamp` entry could likely be just removed, at the cost of needing to do a checkpoint summary lookup when serving the txn via API in order to produce the correct timestamp. Given that data exists elsewhere maybe that's an ok trade off in order to save 8 bytes per txn in the store.

As for the mapping to the checkpoint itself -- That mapping in the perpetual db, and its accessors, is listed as deprecated and "do not use" in a number of places. So my thinking was, given i believe this mapping only exists for the benefit of the the RPC api layer, that a new mapping could live inside of this table and that we could end up removing that map from the perpetual db. This is based off of some conversations i had with @mystenmark in the past so maybe would be good to get his thoughts here too.

Reply:
this mapping is also present in the kv store

Reply:
One other requirement is not relying on kv store for any data :)

Reply:
I guess what I meant is, even if the perpetual table gets completely deprecated, you can do a fanout to local per-epoch store DBs (the provider can set up a retention policy and keep the last N epochs). And if it's not there, we can always query KV